### ISSUE TITLE
Pipelined engine runs at ~36% of the PCIe ceiling: latency-bound, not bandwidth-bound

### ISSUE THREAD

**Issue by pjordanandrsn** [pjordanandrsn/experts4bit-qlora#105]
## The engine is latency-bound, not bandwidth-bound — there is ~2.8x of headroom

Falling out of the #86 analysis, and worth its own issue because it reframes where the next win is.

Qwen3-30B-A3B moves **972 MiB of expert rows per token** (top-8 x 48 layers x 2.53 MiB/expert). Over PCIe4 x16 at a realistic ~20 GB/s that is 51 ms/token, a **19.6 tok/s ceiling**. The engine measures **7.08 tok/s** — about **36% of the link**.

So the cold path is not saturating PCIe. Roughly 2.8x of the ceiling is being lost to something other than bytes.

### This already explains a result we have

The K-ladder measured flat-to-slightly-negative on all three models (see #86's measurement comment: K=-1 7.08, K=8 7.12 on Qwen). If the engine were bandwidth-bound, pinning 32 of 128 experts should have removed ~25% of the routed bytes and shown up as throughput. It did not. That is consistent with bytes not being the constraint — and it means the hot-set dial cannot be tuned into a win no matter what K is, which is the honest reason to stop tuning it.

### Candidates, in the order I would test them

1. **Per-round launch overhead.** The fetch issues one gather launch per round (now up to four, one per segment, after #104). At 48 layers x 1 round/token that is a lot of small launches; the gather itself is a memcpy of 2.53 MiB/expert.
2. **The `have`-skip and dispatch bookkeeping** — `index_select`, `torch.where`, the `row_idx` recompute — all enqueued per fetch, all tiny, all serialized ahead of the copy.
3. **Serialization between the gather and the GEMM.** The GEMM cannot start until the slots are filled; there is no overlap across the k lanes or across layers, so the copy latency is fully exposed rather than pipelined behind compute.
4. **UVA read efficiency.** A host-memory read by address may not reach the same effective bandwidth as a bulk `cudaMemcpyAsync` of the same bytes.

(3) is the one I would look at first: `hint()` already exists as a lead-time prefetch hook and is documented as "mechanism only — untuned". If the copy is exposed latency, prefetching one layer ahead is exactly the shape of fix, and the mechanism is already in the tree.

### How to tell them apart cheaply

Measure the gather in isolation: time N launches of the existing kernel moving the same bytes, and compare against `cudaMemcpyAsync` of an equivalent buffer. That separates "the copy is slow" (4) from "the copy is fine but exposed" (3) without touching the engine.

### Why it matters

The measured payoff of closing even half this gap is larger than anything left in the residency dial: ~7 -> ~13 tok/s on Qwen3-30B-A3B would beat what an NVMe tier could ever deliver (#86's ceiling there was 6.87 tok/s on PCIe4) and costs no extra RAM.

**Comment by pjordanandrsn** [comment]
## Measured on an A5000 — and the answer is not the one the issue proposed

Qwen3-30B-A3B per-layer geometry (E=128, k=8, 2.53 MiB/expert, 20.25 MiB/round), CUDA events, medians over 40+ reps after warmup.

### A. The copy itself is fine — candidate 4 is out

| | time | effective |
|---|---|---|
| address-gather kernel | 1.201 ms | 17.68 GB/s |
| `cudaMemcpyAsync`, same bytes | 1.076 ms | 19.73 GB/s |

The gather runs at **90% of a plain DMA copy**. UVA read efficiency is not the problem.

### B. Where the round goes

```
full decode step           2.366 ms
  fetch (gather + book)    1.420 ms      gather 1.201 (84.6%), bookkeeping 0.219 (15.4%)
  step with slots warm     0.584 ms      gather skips
  fetch + warm-step        2.004 ms      -> 15% of the step is outside BOTH phases
```

48 layers x 2.366 ms = 113.6 ms/token = **8.81 tok/s** on this card. The gather alone is 57.7 ms/token — **51% of the token**, fully serialized ahead of the GEMM.

### C. Candidate 3 is real, but `hint()` cannot fix it

Prefetching through the existing `hint()` hook made things **worse**: −0.222 ms (−8.2%). `hint()` issues the gather on the **same stream**, so it merely queues ahead of the compute and adds a launch. No overlap is available that way.

So I tested whether overlap is available *at all*, on a side stream:

| | serial | overlapped | saved |
|---|---|---|---|
| compute + **gather kernel** | 2.422 ms | 2.432 ms | **−0.4%** |
| compute + **`cudaMemcpyAsync`** | 2.293 ms | 1.464 ms | **+36.2%** |

(ideal = `max(compute, copy)` = 1.222 ms)

**That is the finding.** The address-gather is a *kernel*: it occupies SMs and therefore competes with the GEMM, so it cannot overlap with compute no matter which stream it is on. A DMA copy uses the copy engine and overlaps almost perfectly.

### What this means

The engine sits at 36% of the PCIe ceiling because half the token is an SM-resident copy that is structurally unable to hide behind compute. It also explains the K result cleanly: with the copy on the critical path and not bandwidth-bound, removing bytes by pinning experts was never going to show up.

The recoverable headroom is the ~57.7 ms/token of gather — worth roughly 8.8 -> ~14-18 tok/s if it moves to the DMA engine.

### The hard part, stated honestly

The gather kernel exists for good reasons: arbitrary expert ids -> slots, a device-side have-skip, and **zero host sync** in decode. `cudaMemcpyAsync` needs host-visible addresses, so a naive port reintroduces a per-round host sync — which is likely to cost more than it saves.

Routes worth evaluating, cheapest first:
1. **Copy one step ahead.** Speculative/predicted ids are already available a step early (`hint`'s original purpose). If the ids for round *n+1* are known during round *n*, the addresses are host-visible in time and k `cudaMemcpyAsync` calls can be issued on a copy stream. The have-skip then corrects mispredictions, exactly as `hint` documents.
2. **`cudaMemcpyBatchAsync`** (CUDA 12.8+) for k rows in one call, if it can be driven without a host round-trip per round.
3. **Hybrid**: DMA the predicted rows, keep the gather kernel only for the mispredicted tail.

Not attempting any of these yet — this comment is the measurement, and (1) is the one I would try first since the prediction hook already exists.

**Comment by pjordanandrsn** [comment]
## Measured the DMA route — and it retires most of this issue's premise

I opened this saying the engine runs at ~36% of the PCIe ceiling with ~2.8x of headroom. **That framing was wrong**, and the measurements below are why. Same A5000 setup, Qwen3-30B-A3B per-layer geometry.

### 1. The host sync is not the blocker

The whole reason a DMA route looked hard is that `cudaMemcpyAsync` needs host-visible addresses while decode is zero-host-sync. Measured on a *loaded* queue (the honest case — a sync is cheap on an idle stream and expensive when it drains real work):

| | |
|---|---|
| `.cpu()` on an idle stream | 0.063 ms |
| gather + `.cpu()` | 2.050 ms |
| gather alone | 1.982 ms |
| **sync on a loaded queue** | **+0.068 ms** |

~68 us against a ~2 ms layer. Cheap. I had assumed this would sink the idea; it does not.

### 2. DMA moves the same bytes faster than the gather

| | | |
|---|---|---|
| gather kernel (today) | 2.013 ms | 10.55 GB/s |
| k separate DMA copies | 1.771 ms | 11.99 GB/s |
| one bulk DMA copy | 1.755 ms | 12.10 GB/s |

k scattered copies land within 1% of one bulk copy, so per-copy overhead is not a problem — scattered per-expert DMA is viable.

### 3. But the pipeline idea is actively harmful

I proposed copying expert i+1 while the GEMM runs expert i, since that is the only overlap available without a predictor (within a layer the GEMM consumes what the copy produces, so they cannot overlap for the same expert).

| | | |
|---|---|---|
| A. today: gather + fused GEMM | 2.193 ms | — |
| B. k DMA copies + fused GEMM | 1.966 ms | **+10.4%** |
| C. k-way pipeline, copy i+1 \|\| GEMM i | 3.240 ms | **−47.8%** |

C is far worse. Splitting the fused GEMM into k tiny launches costs much more than the overlap recovers — which is exactly the property the engine was designed around, and I proposed giving it up without checking.

### What is actually recoverable

**~7-10%**, not 2x: swap the gather for DMA copies and keep the fused GEMM (B), minus the 0.068 ms sync — so about +7% net, at the cost of the zero-host-sync decode contract that `test_zero_sync_token_loop` enforces.

The "2.8x headroom" never existed. It assumed the copy could be hidden behind compute, and within a layer there is nothing to hide it behind: attention precedes the router, the router precedes the ids, and the GEMM depends on the copy. The copy is irreducible work on the critical path, and DMA is only ~13% faster at doing it. On this box the copy path reaches ~12 GB/s, so 972 MiB/token is ~81 ms of pure transfer against ~105 ms measured — the engine is at roughly **77%** of its copy-bound floor, not 36%.

### Recommendation

Not worth building as scoped. A 7% gain that trades away the zero-host-sync property and adds a per-layer `.cpu()` is a bad deal for an engine whose central design claim is that no CPU knows the ids.

Worth revisiting only if a cheap id predictor appears (then the copy moves off the critical path entirely and the arithmetic changes completely), or on hardware where PCIe is much faster relative to compute. Leaving this open as a record of the measurement rather than as a work item.

**Comment by pjordanandrsn** [comment]
## What CUDA-graph capture is worth — the number that decides the DMA swap

I argued above that swapping the gather for DMA is a bad trade because it needs a per-layer `.cpu()` and would break capture, which is already implemented and tested (`tests/test_pipelined_graphs.py`). That argument only holds if capture is worth more than the ~7% the swap buys. Measured on an A4000, Qwen3-30B-A3B per-layer geometry:

| | GPU ms | wall ms |
|---|---|---|
| eager (ids churning) | 2.182 | 0.414 |
| graph replay | 2.051 | **0.012** |
| **replay saves** | **+6.0%** | **+97.0%** |

**Correction to my own first run:** I initially measured +37.6% GPU. That was wrong — the benchmark reused one id set and never reset `have`, so after warmup every lane took the skip and it timed a GEMM-only step with no copy at all. With ids churning like a real token stream, capture is worth **6.0%** of GPU time, not 37.6%.

### So the two options are a wash on GPU time

| | GPU | other effects |
|---|---|---|
| DMA swap | ~+7% | frees SMs; **breaks capture**; new code |
| graph capture | +6.0% | **−97% CPU time per layer**; already built and tested |

Capture only removes launch overhead, and the step is ~87% copy — so it cannot do better than this, exactly as the copy-bound picture predicts.

### Recommendation stands, for a weaker reason than I gave

Keep the gather — but not because capture dominates. It doesn't; they're within a point of each other on GPU time. The reason is that **capture already exists, costs nothing to keep, and additionally drops per-layer CPU time from 0.414 ms to 0.012 ms**, which is what starts to matter under batched or multi-stream serving where the enqueue side becomes the limit. DMA would be new code that removes a tested capability for a GPU-time wash.

If someone later measures the CPU side as genuinely irrelevant for the target deployment, DMA becomes a legitimate ~7% and this should be revisited. Recording that condition explicitly so the decision can be reopened on evidence rather than re-argued.

**Comment by pjordanandrsn** [comment]
## Closing: the premise in the title is retracted

The "~36% of the PCIe ceiling / ~2.8x of headroom" framing was **wrong**, and the error was in the ceiling, not the measurement.

**The ceiling came from a spec sheet, not the box.** I used ~20 GB/s for PCIe4 x16. These A5000 hosts actually reach **~10–12 GB/s** measured. That single wrong constant is most of the phantom headroom — the engine was never 2.8x off its link.

**The copy is not slow, it is unhideable.** Measured with CUDA events, medians over 40+ reps, on Qwen3-30B-A3B per-layer geometry:

| | serial | on a side stream | saved |
|---|---|---|---|
| compute + gather kernel | 2.422 ms | 2.432 ms | **−0.4%** |
| compute + `cudaMemcpyAsync` | 2.293 ms | 1.464 ms | **+36.2%** |

The address-gather is a **kernel**: it occupies SMs and competes with the GEMM, so it cannot overlap compute on any stream. A DMA copy uses the copy engine and overlaps nearly perfectly. That is the real mechanism — not latency in the abstract.

But there is nothing to overlap it *with*. Inside a layer, attention precedes the router, the router precedes the ids, and the GEMM consumes what the copy produces, so the expert copy is irreducible work on the critical path. Testing the DMA route end-to-end:

- host sync on a loaded queue is only **+0.068 ms** — it was never the blocker
- DMA moves the same bytes ~13% faster than the gather (12.10 vs 10.55 GB/s)
- gather + fused GEMM 2.193 ms → k DMA copies + fused GEMM 1.966 ms (**+10.4%**, ~+7% after the sync)
- k-way pipelining (copy i+1 while GEMM i runs) is **−47.8%** — splitting the fused GEMM into k tiny launches costs far more than the overlap returns

**So the engine is at ~77% of its copy-bound floor, not 36%.** Roughly +7% is available, and it costs the zero-host-sync decode contract. Not worth trading. Revisit only with a cheap id predictor, which moves the copy off the critical path and changes the arithmetic entirely.

### What the candidate list was right about

Candidate 1, per-round launch overhead, was real and is now shipped. Launches were **288.8 ms of 562.4 ms host time across 45,501 launches** (OLMoE/3090). CUDA-graph capture collapses that to one launch per step — #110, released in 0.16.2. The honest measurement there: **4.4–5.6x on tiny fixtures, 1.11x on OLMoE-1B-7B, 1.04x on Qwen3-30B-A3B**, i.e. inversely proportional to real GPU work per step. Launch overhead was a genuine cost and removing it is genuinely small at scale.

The K-ladder conclusion in this issue also stands, but for a different reason than stated — see #108.

### Method note, since this cost two wrong answers

A "ceiling" computed from a datasheet is a **model**, and `max(copy, compute)` silently assumes an overlap the dependency graph forbids. Measure the candidate design end-to-end instead: mine came out 47.8% slower than the thing it was supposed to beat.
